### DISCUSSION TITLE
How to: Scoped helper functions

### DISCUSSION THREAD

**Discussion by okaufmann** [pestphp/pest#495] [Q&A]
Hi guys

In complex test scenarios I'd like to have small helper methods like before having private methods in PHPUnit test classes.

This works fine, as long these functions do not have the same name.

Is there any way having a `createUser()` method in more than one Pest test file?

Otherwise, how could I achieve this similarly?

Here is an example of what I mean (I'm aware that this is a use case where we can use Factories, but there may be other and different things per test file I'd like to do in `createUser` method)

```php
<?php

// PostTest.php

it('can create a Post', function() {
    // ...
});

function createUser($param1, $param2) {
    // ...
}
```

```php
<?php

// CommentTest.php

it('can create a Comment', function() {
    // ...
});

function createUser($param3, $param4, $param5) {
    // ...
}
```

Help appreciated 🙏

**Comment by okaufmann** [comment]
After some Tinkering I solved this by using namespaces to prevent name conflicts.

I just include the filename in the namespace declaration to "trick" the autoloading.


```php
<?php

// tests/Feature/PostTest.php

namespace Tests\Feature\PostTest;

function createUser()
{
    // ...
}

it('can create a Post', function() {
    // ...
});
```


```php
<?php

// tests/Feature/CommentTest.php

namespace Tests\Feature\CommentTest;

function createUser()
{
    // ...
}

it('can create a Comment', function() {
    // ...
});
```

As there are no classes within these files, I think this is OK and does not hurt the PSR-4 rule.
Also the methods can be identified by intellisense correctly.

<img width="754" alt="CleanShot 2022-03-07 at 23 33 31@2x" src="https://user-images.githubusercontent.com/4414498/157129348-887b9444-d9c5-488f-8d83-1aab979cb725.png">

What do you guys think?

**Reply by owenvoke** [reply]
That's what I tend to do, or a trait. But it wouldn't work in your case as the methods are different.

**Reply by okaufmann** [reply]
Yes, a Trait was my first try, but I like to keep them together in the same test file.

**Reply by j-dexx** [reply]
I think it'd be nice to have a specific way to do this rather than having a namespace.
